core/src/test/java/org/elasticsearch/discovery/zen/UnicastZenPingTests.java

```diff
@@ -252,6 +254,151 @@ public ClusterState clusterState() {
         assertCounters(handleD, handleA, handleB, handleC, handleD);
     }
 
+    public void testHostProviderBuildDynamicNodesShouldBeCalledOnlyOnceInEachRound() throws IOException, InterruptedException {
+        final Settings settings = Settings.builder().put("cluster.name", "test").put(TransportSettings.PORT.getKey(), 0).build();
+
+        NetworkService networkService = new NetworkService(settings, Collections.emptyList());
+
+        final BiFunction<Settings, Version, Transport> supplier = (s, v) -> new MockTcpTransport(
+            s,
+            threadPool,
+            BigArrays.NON_RECYCLING_INSTANCE,
+            new NoneCircuitBreakerService(),
+            new NamedWriteableRegistry(Collections.emptyList()),
+            networkService,
+            v);
+
+        NetworkHandle handleA = startServices(settings, threadPool, "UZP_A", Version.CURRENT, supplier);
+        closeables.push(handleA.transportService);
+
+        final ClusterState state = ClusterState.builder(new ClusterName("test")).version(randomPositiveLong()).build();
+
+        Settings hostsSettings = Settings.builder()
+            .putArray("discovery.zen.ping.unicast.hosts",
+                NetworkAddress.format(new InetSocketAddress(handleA.address.address().getAddress(), handleA.address.address().getPort())))
+            .put("cluster.name", "test")
+            .build();
+
+        UnicastHostsProvider unicastHostsProviderA = mock(UnicastHostsProvider.class);
+        UnicastZenPing zenPingA = new UnicastZenPing(hostsSettings, threadPool, handleA.transportService, unicastHostsProviderA);
+        zenPingA.start(new PingContextProvider() {
+            @Override
+            public DiscoveryNodes nodes() {
+                return DiscoveryNodes.builder().add(handleA.node).localNodeId("UZP_A").build();
+            }
+
+            @Override
+            public ClusterState clusterState() {
+                return ClusterState.builder(state).blocks(ClusterBlocks.builder().addGlobalBlock(STATE_NOT_RECOVERED_BLOCK)).build();
+            }
+        });
+        closeables.push(zenPingA);
+
+        zenPingA.pingAndWait(TimeValue.timeValueSeconds(1));
+        verify(unicastHostsProviderA, times(1)).buildDynamicNodes();
+    }
+
+    public void testShouldPingDiscoveredNodes() throws IOException, InterruptedException {
```

Comment:
can we call this `testDiscoveryOfNoneConfiguredNodes` and add a comment: `test that nodes discover each other if they ping a common host` ?

This doesn't test what I intended - i.e., test that a node pings nodes that have pinged it while it was pinging. That is however non trivial to test and I don't want to delay your change for setting this test up.